### DISCUSSION TITLE
Add in the ability to customize the copy semantics for structure types.

### DISCUSSION THREAD

**Discussion by kkukshtel** [dotnet/csharplang#7556] [Language Ideas]
Hello all!

I was recently playing with the new `record struct` type and noticed that in the docs on [Custom Copy Semantics](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/with-expression#custom-copy-semantics) it mentions that "You can't customize the copy semantics for structure types."

Is there any specific reason for this? Is there anyway around it? I'd really like to do something like this:

```c#
//define an entity
var id_counter = 0;
public record struct Entity(unit ID) {}
{
    public Entity(Entity original)
    {
        ID = id_counter++;
    }
}

var a = Entity {} //ID resolves to 1;
var b = a with {}; //ID resolves to 2
```

Instead right now, the copy semantic is ignored, not used, and both `a` and `b` have the same ID.

Notably this _does_ work if `Entity` is instead a `record class`/`record`.

Thanks!

**Comment by HaloFour** [comment]
Structs are copied by value, they don't require a clone method like classes do.  When you use `with` on a record struct, all it does is assign to another local, which automatically creates a copy, and then mutates that copy.

**Reply by kkukshtel** [reply]
Right, but the goal is to have it actually call the Copy ctor on `with`, at least for what I'm looking for. As you say right now it ignores it.

**Reply by kkukshtel** [reply]
Somone on the discord as well brought up that something like a finalizer would also suit a bit

**Comment by 333fred** [comment]
As noted on discord, here's the meeting notes where this was decided: https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-09-30.md#with-expressions.

**Reply by kkukshtel** [reply]
It was also brought up on the discord the specific note in those notes that "we could design a new runtime intrinsic to call either dup or the struct's clone method if it exists". Just wondering if this decision has been revisisted.

**Reply by 333fred** [reply]
As I mentioned in discord, no.

**Comment by msedi** [comment]
I think there have been so many requests about the struct topic, that some decisions should be revised. Especially copy semantics and destructors are often requested. Sure, this topic might be complex, but would solve a lot of problems.
